NAS-125354 / 24.04 / Added ixAddNewOption directive

RehanY147 opened the pull request:
There are two examples.
The Replication Add Wizard -> Advanced Options -> SSH Connection (under Transport Options)
The Add Cloud Sync Task -> Credential (under Remote)

On the slide-in form for Cloud Sync tasks, the credentials select should have the option 'Add New'. When you click it, the form should open to add `Cloud Credentials`. Once successfully added a cloud credential, the cloud credential form should close and UI should go back to the Cloud Sync Task card. The provider select should have the newly created value as the selected value.

bugclerk commented:
Jira URL: https://ixsystems.atlassian.net/browse/NAS-125354

undsoft reviewed (changes requested):
Looks promising, just curious if we could do something in the directive itself to listen to what new item was added and add it to the list of options, so that there is less code outside of the component to handle this.
We could also require forms to implement a certain interface to make sure they return something standard when they are closed.

RehanY147 commented:
@undsoft We don't know what the value and label is every time. It's not the same attribute every time. That's the main problem with trying to add the option in the directive.

denysbutenko commented on src/app/pages/data-protection/cloudsync/cloudsync-form/cloudsync-form.component.ts:
```diff
@@ -123,7 +127,8 @@ export class CloudsyncFormComponent implements OnInit {
 
   credentialsList: CloudsyncCredential[] = [];
-  readonly credentialsOptions$ = this.cloudCredentialService.getCloudsyncCredentials().pipe(
+  readonly fetchCredentialsOptions$ = new Subject<void>();
```
Unused

denysbutenko commented on src/app/pages/data-protection/cloudsync/cloudsync-form/cloudsync-form.component.ts:
```diff
@@ -67,6 +69,8 @@ export class CloudsyncFormComponent implements OnInit {
       : this.translate.instant('Edit Cloud Sync Task');
   }
 
+  CloudCredentialsFormComponent = CloudCredentialsFormComponent;
```
That's weird. Maybe we can follow other approaches. For example, what if we can do a map of possible options and components and then use an enum to avoid importing the whole component like that?

denysbutenko commented on src/app/pages/data-protection/cloudsync/cloudsync-form/cloudsync-form.component.ts:
```diff
@@ -180,8 +184,24 @@ export class CloudsyncFormComponent implements OnInit {
   ) { }
 
+  newCloudCredAdded(value: unknown): void {
+    this.credentialsOptions$ = this.cloudCredentialService.getCloudsyncCredentials().pipe(
```
duplicate at line [#131](https://github.com/truenas/webui/pull/9347/files#diff-07f66c4dfca3295b93b45129b462776611624e53d3e6494943c35260eee67dbaR131). lets merge them

denysbutenko commented on src/app/pages/data-protection/cloudsync/cloudsync-task-card/cloudsync-task-card.component.ts:
```diff
@@ -51,6 +51,8 @@ export class CloudSyncTaskCardComponent implements OnInit {
   jobStates = new Map<number, string>();
   readonly jobState = JobState;
 
+  private ixChainedSlideInService = inject(IxChainedSlideInService);
```
I like this approach, but it breaks component consistency. Could we move it to the constructor?

denysbutenko commented on src/app/modules/ix-forms/components/ix-select/ix-add-new-option.directive.ts:
```diff
@@ -0,0 +1,58 @@
+          this.chainedSlideIn.pushComponent({ component: this.ixAddNewOption, close$ });
+          close$.pipe(take(1), untilDestroyed(this)).subscribe({
+            next: (response) => {
+              this.newOptionAdded.emit(response);
```
Do you think we can update select value and options from here? It may allow us to eliminate callbacks.

denysbutenko reviewed (changes requested):
It works okay and needs some improvements.

undsoft commented on src/app/modules/ix-forms/components/ix-select/ix-select-with-new-option.directive.ts:
```diff
@@ -0,0 +1,125 @@
+  abstract fetchUpdatedOptions(): Observable<Option[]>;
+
+  ngAfterViewInit(): void {
+    if (this.ixSelect) {
```
return early

undsoft commented on src/app/modules/ix-forms/components/ix-select/ix-select-with-new-option.directive.ts:
```diff
@@ -0,0 +1,125 @@
+  private translateService = inject(TranslateService);
+
+  ngOnInit(): void {
+    this.fetchUpdatedOptions().pipe(
```
Maybe just call it `fetchOptions`.

undsoft reviewed (approved):
An interesting approach with abstract class. Looks nice.

Don't forget to remove Create New option from Replication form.

RehanY147 commented:
@undsoft @denysbutenko This PR has also reworked the `cloudsync-wizard` to use the ChaindSlideInService and also added a swap slide-in functionality. Test
- The component that opened the original wizard/form should still get the response when the form is submitted even if you switch to form from wizard or vise-versa while using the slide-in.
- The CloudsyncWizard component should work as before without any regressions.
- Test the add new option and the abstract class that is used to create new ix-selects with add-new option

undsoft reviewed (changes requested):
Cloudsync wizard - when I add new item, I cannot advance to next step. Next button doesn't do anything.
Rsync task form - maybe unrelated to this PR, but I couldn't get Save button to be enabled.

RehanY147 commented:
@undsoft It's working for me.

https://github.com/truenas/webui/assets/24204846/1517bf22-0e2e-4faf-8a7c-5bdccb575c84

undsoft reviewed (approved):
Yeah, weird. Appears to be working okay now.

denysbutenko reviewed (changes requested):
I have noticed a few regressions:
- Editing cloud sync task shows Credentials select empty.
- Credential options are not sorted anymore.
- I see an empty slide-in panel when I go to `/credentials/backup-credentials` and click on the edit button cloud credential. Error in console `No provider for InjectionToken CHAINED_SLIDE_IN_REF!`

RehanY147 commented:
@denysbutenko I don't think sorting the options was an intentional behavior. Now they are just sorted the other way around
<img width="397" alt="image" src="https://github.com/truenas/webui/assets/24204846/e4e92663-574e-499e-a780-78b51125776d">

But good catch on the CloudCredentialsForm! Thanks!

denysbutenko commented on src/app/modules/ix-forms/components/ix-slide-in/components/ix-chained-slide-in/ix-chained-slide-in.component.ts:
```diff
@@ -14,12 +16,10 @@ import { ChainedComponentSeralized as ChainedComponentInfoSeralized, IxChainedSl
   @Input() id: string;
   @ViewChild('componentWrapper') container: HTMLElement;
-  protected components: ChainedComponentInfoSeralized[];
+  protected components: ChainedComponentSeralized[];
```
Typo in `seralized`

denysbutenko reviewed (approved):
Looks good, great work 👍🏻

bugclerk commented:
This PR has been merged and conversations have been locked.
If you would like to discuss more about this issue please use our forums or raise a Jira ticket.
